authui/src/authflowv2/loading.ts

```diff
@@ -0,0 +1,63 @@
+import { Controller } from "@hotwired/stimulus";
+import {
+  disableAllButtons,
+  hideProgressBar,
+  showProgressBar,
+} from "../loading";
+
+interface LoadingEvent extends Event {
+  detail?: {
+    button?: HTMLButtonElement;
+    error?: unknown;
+  };
+}
+
+export class LoadingController extends Controller {
+  revertDisabledButtons: { (): void } | null = null;
+  button: HTMLButtonElement | null = null;
+
+  // Revert disabled buttons before Turbo caches the page
+  // To avoid flickering in the UI
+  beforeCache = () => {
+    if (this.revertDisabledButtons) {
+      this.revertDisabledButtons();
+    }
+  };
+
+  onLoading(e: LoadingEvent) {
+    this.revertDisabledButtons = disableAllButtons();
+    showProgressBar();
+    if (e.detail?.button?.getAttribute("data-loading-state") != null) {
+      e.detail.button.classList.add("primary-btn--loading");
+      this.button = e.detail.button;
+    }
+  }
+  onLoadingError(e: LoadingEvent) {
+    // revert is only called for error branch because
+    // The success branch also loads a new page.
+    // Keeping the buttons in disabled state reduce flickering in the UI.
+    if (this.revertDisabledButtons) {
+      this.revertDisabledButtons();
+      this.revertDisabledButtons = null;
+    }
+    this.button?.classList.remove("primary-btn--loading");
```

Comment:
Same as above, use a generic attribute instead of class coupled with primary button.